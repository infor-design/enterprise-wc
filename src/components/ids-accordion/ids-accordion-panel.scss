@import '../../core/ids-base';

// ============================================
// Top-level App Menu Accordion Panel Styles
@mixin accordion-panel-app-menu-colors() {
  @include mb-4();

  transition:
    background-color $transition-length $transition-ease,
    height $transition-length $transition-ease,
    padding $transition-length $transition-ease;

  &.expanded {
    @include py-8();
  }

  // Light Mode Styles
  &[mode='light'],
  &:not([mode]) {
    &.expanded {
      @include bg-black();
    }
  }

  &[mode='dark'] {
    &.expanded {
      @include bg-slate-100();
    }
  }

  &[mode='contrast'] {
    &.expanded {
      @include bg-slate-100();
    }
  }
}

// ============================================
// Sub-level App Menu Accordion Panel Styles
@mixin accordion-panel-sub-menu-colors() {
  @include m-0();
}

// ============================================
// Controls the size of nested accordions
@mixin accordion-pane-standard-size() {
  .ids-accordion-pane-content {
    @include px-8();
  }

  &:not(.rtl) {
    &.nested {
      .ids-accordion-pane-content {
        @include pl-16();
        @include pr-0();
      }
    }
  }

  // RTL Styles on Nested Accordions
  // (applies the padding to the opposite side)
  &.rtl {
    &.nested {
      .ids-accordion-pane-content {
        @include pl-0();
        @include pr-16();
      }
    }
  }
}

@mixin accordion-pane-nested-size() {
  // LTR nested styles (default)
  &:not(.rtl) {
    .ids-accordion-pane-content {
      @include pl-28();
    }
  }

  // RTL nested styles
  &.rtl {
    .ids-accordion-pane-content {
      @include pr-28();
    }
  }
}

:host {
  display: block;
}

.ids-accordion-panel {
  // ============================================
  // Color Variant Rules
  &:not(.color-variant-app-menu):not(.color-variant-sub-app-menu) {
    @include accordion-pane-standard-size();
  }

  &.color-variant-app-menu {
    @include accordion-panel-app-menu-colors();
    @include accordion-pane-standard-size();
  }

  &.color-variant-sub-app-menu {
    @include accordion-panel-sub-menu-colors();
    @include accordion-pane-nested-size();
  }

  // ============================================
  // Content Alignment Rules

  &.has-icon {
    // LTR content alignment (padding comes from the left)
    &:not(.rtl) {
      ::slotted(ids-accordion-panel) {
        @include pl-24();
      }

      .ids-accordion-panel {
        @include pl-24();
      }
    }

    // RTL content alignment (padding comes from the right)
    &.rtl {
      ::slotted(ids-accordion-panel) {
        @include pr-24();
      }

      .ids-accordion-panel {
        @include pr-24();
      }
    }
  }
}

.ids-accordion-pane {
  @include box-border();

  contain: content;
  height: 0;
  overflow: hidden;
  transition: height $transition-length $transition-ease;
}

.ids-accordion-pane-title {
  @include audible();
}
