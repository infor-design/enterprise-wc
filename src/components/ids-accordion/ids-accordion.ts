import { customElement, scss } from '../../core/ids-decorators';
import { attributes } from '../../core/ids-attributes';
import { stringToBool } from '../../utils/ids-string-utils/ids-string-utils';
import Base from './ids-accordion-base';
import './ids-accordion-header';
import './ids-accordion-panel';
import styles from './ids-accordion.scss';

/**
 * IDS Accordion Component
 * @type {IdsAccordion}
 * @inherits IdsElement
 * @mixes IdsColorVariantMixin
 * @mixes IdsEventsMixin
 * @mixes IdsKeyboardMixin
 * @mixes IdsLocaleMixin
 * @mixes IdsThemeMixin
 * @part accordion - the accordion root element
 */
@customElement('ids-accordion')
@scss(styles)
export default class IdsAccordion extends Base {
  constructor() {
    super();
  }

  connectedCallback() {
    super.connectedCallback?.();
    this.#handleEvents();
    this.#handleKeys();

    // Assign depth-dependent styles, and re-apply them on changes
    requestAnimationFrame(() => {
      this.#assignDepthDependentStyles();
      this.#contentObserver.observe((this as any), {
        childList: true
      });
    });
  }

  /**
   * Return the attributes we handle as getters/setters
   * @returns {Array} The attributes in an array
   */
  static get attributes() {
    return [
      ...super.attributes,
      attributes.ALLOW_ONE_PANE,
      attributes.DISABLED,
      attributes.MODE,
      attributes.VERSION
    ];
  }

  /**
   * @returns {Array<string>} List of available color variants for this component
   */
  colorVariants = ['app-menu'];

  /**
   * When the accordion's color variant is set, push this change through to the child elements
   */
  onColorVariantRefresh() {
    this.#assignDepthDependentStyles(this, 0, true, false, false, false);
  }

  /**
   * Inner template contents
   * @returns {string} The template
   */
  template() {
    return `
      <div class="ids-accordion" part="accordion">
        <slot></slot>
      </div>
    `;
  }

  /**
   * Observes changes in the accordion tree
   */
  #contentObserver = new MutationObserver((mutations) => {
    for (const m of mutations) {
      if (m.type === 'childList') {
        this.#assignDepthDependentStyles();
      }
    }
  });

  /**
   * @readonly
   * @returns {Array<HTMLElement>} all accordion headers in a flattened array
   */
  get headers(): Array<HTMLElement> {
    return [...this.querySelectorAll('ids-accordion-header')];
  }

  /**
   * @readonly
   * @returns {Array<any>} all accordion panels in a flattened array
   */
  get panels(): Array<any> {
    return [...this.querySelectorAll('ids-accordion-panel')];
  }

  /**
   * @readonly
   * @returns {any} the currently focused menu item, if one exists
   */
  get focused(): any {
    if (this.contains(document.activeElement)) {
      return (document.activeElement as any).closest('ids-accordion-panel');
    }
    return undefined;
  }

  /**
   * Gets allowOnePane property
   * @readonly
   * @returns {boolean} true if accordion set to allowOnePane
   */
  get allowOnePane(): boolean {
    return stringToBool(this.getAttribute(attributes.ALLOW_ONE_PANE));
  }

  /**
   * Sets allowOnePane property
   * @param {boolean|string} allow true/false
   */
  set allowOnePane(allow: boolean | string) {
    const toAllow = stringToBool(allow);

    if (toAllow) {
      this.setAttribute(attributes.ALLOW_ONE_PANE, `${toAllow}`);
    } else {
      this.removeAttribute(attributes.ALLOW_ONE_PANE);
    }
  }

  /**
   * Gets disabled property
   * @readonly
   * @returns {boolean} true if accordion set to disable
   */
  get disabled() {
    return stringToBool(this.getAttribute(attributes.DISABLED));
  }

  /**
   * Sets disabled property
   * @param {boolean|string} value true/false
   */
  set disabled(value) {
    const disabled = stringToBool(value);

    if (disabled) {
      this.setAttribute(attributes.DISABLED, `${disabled}`);
    } else {
      this.removeAttribute(attributes.DISABLED);
    }

    this.panels.forEach((panel) => {
      panel.disabled = disabled;
    });
  }

  /**
   * Labels Headers and Panels with styling information that is
   * dependent on how deeply-nested they are within the Accordion tree.
   * @param {HTMLElement} element the element to check
   * @param {number} depth the zero.
   * @param {boolean} doColorVariant if true, modifies the color variant
   * @param {boolean} doExpanderType if true, modifies the expander type
   * @param {boolean} doDisplayIconType if true, modifies the display icon type
   * @param {boolean} doRTL if true, modifies RTL styles
   */
  #assignDepthDependentStyles(
    element = this,
    depth = 0,
    doColorVariant = true,
    doExpanderType = true,
    doDisplayIconType = true,
    doRTL = true
  ) {
    this.header = element.querySelector(':scope > ids-accordion-header');
    const subLevelDepth = depth > 1;

    if (depth > 0) {
      // Assign Nested Padding CSS Classes
      element.nested = subLevelDepth;

      // Assign Color Variant
      if (doColorVariant && this.colorVariant) {
        const variant = subLevelDepth ? `sub-${this.colorVariant}` : this.colorVariant;
        element.colorVariant = variant;

        if (this.header) {
          this.header.colorVariant = variant;
        }
      }

      if (this.header) {
        // Pass language/locale down to child components
        this.header.language = this.language?.name;

        // Assign Expander Type
        // (Use Plus/Minus-style expander on any nested panels)
        if (doExpanderType) {
          const expanderType = subLevelDepth ? 'plus-minus' : 'caret';
          this.header.expanderType = expanderType;
        }

        // Assign Content Alignment Style
        // (applies special alignment rules to ALL panes
        // adjacent to panes containing an icon in their header)
        if (doDisplayIconType) {
          const displayIconType = this.header.icon;
          if (typeof displayIconType === 'string' && displayIconType.length && !element.contentAlignment) {
            this.#markAdjacentPanesForIcons(element, true);
          }
        }
      }
    }

    // Check children for nested panes
    const children = element.children;
    for (const childEl of children) {
      if (depth > 6) {
        break;
      }
      if (childEl.tagName !== 'IDS-ACCORDION-PANEL') {
        continue;
      }
      this.#assignDepthDependentStyles(
        childEl,
        depth + 1,
        doColorVariant,
        doExpanderType,
        doDisplayIconType,
        doRTL
      );
    }
  }

  /**
   * @returns {void}
   */
  #handleEvents() {
    this.offEvent('languagechange.accordion-container');
    this.onEvent('languagechange.accordion-container', this.closest('ids-container'), (e: CustomEvent) => {
      if (this.header) {
        this.header.language = e.detail.language.name;
      }
      this.#assignDepthDependentStyles(this, 0, false, false, false, true);
    });

    // Responds to `selected` events triggered by children
    this.onEvent('selected', this, (e: CustomEvent) => {
      this.#deselectOtherHeaders((e.target as HTMLElement));
    });
  }

  /**
   * Makes accordion headers appear to be deselected, except for the provided one.
   * @param {HTMLElement} target a header to ignore
   */
  #deselectOtherHeaders(target: HTMLElement) {
    this.headers.forEach((header: any) => {
      if (header.selected && !target.isEqualNode(header)) {
        header.selected = false;
      }
    });
  }

  /**
   * Sets up keyboard navigation among accordion elements
   * @returns {void}
   */
  #handleKeys() {
    // Arrow Up navigates focus backward
    this.listen(['ArrowUp'], this, (e: KeyboardEvent) => {
      e.preventDefault();
      e.stopPropagation();
      this.#prevPanel();
    });

    // Arrow Down navigates focus forward
    this.listen(['ArrowDown'], this, (e: KeyboardEvent) => {
      e.preventDefault();
      e.stopPropagation();
      this.#nextPanel();
    });
  }

  /**
   * Traverses the Accordion a specified number of steps
   * @param {number} amt the amount of steps to take
   * @returns {HTMLElement} the newly-focused accordion pane
   */
  navigate(amt = 0) {
    if (typeof amt !== 'number') {
      return this.focused;
    }

    const negative = amt < 0;
    let steps = Math.abs(amt);
    while (steps > 0) {
      if (negative) {
        this.#prevPanel();
      } else {
        this.#nextPanel();
      }
      steps -= 1;
    }
    return this.focused;
  }

  /**
   * Navigates focus from the currently focused Accordion Panel to the next,
   * looping focus to the first panel if applicable.
   * @returns {void}
   */
  #nextPanel() {
    const currentItem = this.focused;
    let next;

    // If the focused panel is expandable, find the first panel inside of it
    if (currentItem.isExpandable && currentItem.expanded) {
      next = currentItem.querySelector('ids-accordion-panel') || currentItem.nextElementSibling;
    } else {
      next = currentItem.nextElementSibling;
    }

    // If there's no next sibiling, or this pane has been closed,
    // navigate to next item outside this panel
    if (!next) {
      next = currentItem.parentElement.nextElementSibling;
    }

    // If next is not an accordion panel, consider that we've 'looped'
    // back around to the top and pick the first panel
    if (!next || next.tagName !== 'IDS-ACCORDION-PANEL') {
      next = this.querySelector('ids-accordion-panel');
    }

    next.focus();
  }

  /**
   * Navigates focus from the currently focused Accordion Panel to the previous,
   * looping focus to the last panel if applicable.
   * @returns {void}
   */
  #prevPanel() {
    const currentItem = this.focused;
    const getLastPanel = () => {
      const prevChildren = currentItem.parentElement.querySelectorAll('ids-accordion-panel:last-child');
      return prevChildren[prevChildren.length - 1];
    };

    let prev = currentItem.previousElementSibling;
    if (!prev) {
      prev = getLastPanel();
    }

    // If the previous panel is expandable, focus on its last pane instead
    if (prev.isExpandable && prev.expanded) {
      const current = prev;
      prev = prev.querySelector('ids-accordion-panel:last-child') || current;
    }

    // If the previous element is a header, no more panels are present.
    // Navigation should be pushed one panel level up;
    if (prev.tagName === 'IDS-ACCORDION-HEADER') {
      prev = prev.parentElement;
    }

    // If this pane has been closed, navigate to previous item outside this pane
    while (prev.parentElement.tagName === 'IDS-ACCORDION-PANEL' && !prev.parentElement.expanded) {
      prev = prev.parentElement;
    }

    // If there's no previous sibiling, navigate to the previous highest pane
    if (!prev) {
      prev = currentItem.parentElement;
    }

    // If previous is not an accordion panel, consider that we've 'looped'
    // back around to the top and pick the first header
    if (!prev || prev.tagName !== 'IDS-ACCORDION-PANEL') {
      prev = getLastPanel();
    }

    prev.focus();
  }

  /**
   * Assigns CSS classes to panes/headers that will correctly align their contents
   * for either having an icon, or not having an icon
   * @param {any} panel the accordion panel that contains icons
   * @param {boolean} status true if other adjacent accordion panels
   *   should appear to be aligned with this panel's icon
   */
  #markAdjacentPanesForIcons(panel: any, status: boolean) {
    const parent = panel.parentElement;
    [...parent.children].forEach((node) => {
      if (node.tagName === 'IDS-ACCORDION-PANEL') {
        node.contentAlignment = status ? 'has-icon' : null;
      }
    });
  }
}
