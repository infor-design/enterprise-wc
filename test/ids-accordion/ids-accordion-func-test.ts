/**
 * @jest-environment jsdom
 */
import '../../src/components/ids-accordion/ids-accordion';
import IdsAccordionHeader from '../../src/components/ids-accordion/ids-accordion-header';
import IdsContainer from '../../src/components/ids-container/ids-container';
import waitFor from '../helpers/wait-for';

import createFromTemplate from '../helpers/create-from-template';

const createAccordion = async (accordion: any, variant?: string | null) => {
  const variantProp = variant ? ` color-variant="${variant}"` : '';
  accordion = await createFromTemplate(accordion, `<ids-accordion${variantProp}>
  <ids-accordion-panel id="p1">
    <ids-accordion-header id="h1" slot="header">
      <ids-text>Header 1</ids-text>
    </ids-accordion-header>
    <ids-text slot="content">content</ids-text>
  </ids-accordion-panel>
  <ids-accordion-panel id="p2">
    <ids-accordion-header id="h2" slot="header">
      <ids-text>Header 2</ids-text>
    </ids-accordion-header>
  </ids-accordion-panel>
  <ids-accordion-panel id="p3">
    <ids-accordion-header id="h3" slot="header">
      <ids-text>Header 3</ids-text>
    </ids-accordion-header>
  </ids-accordion-panel>
</ids-accordion>`);
  return accordion;
};

describe('IdsAccordion Component', () => {
  let accordion: any;
  let panel: any;
  let panel2: any;
  let panel3: any;
  let header: any;
  let header2: any;

  beforeEach(async () => {
    accordion = await createAccordion(accordion);

    panel = document.querySelector('#p1');
    panel2 = document.querySelector('#p2');
    panel3 = document.querySelector('#p3');
    header = (document.querySelector('#h1') as unknown as IdsAccordionHeader);
    header2 = (document.querySelector('#h2') as unknown as IdsAccordionHeader);
  });

  afterEach(async () => {
    accordion.remove();
    accordion = null;
    panel = null;
    panel2 = null;
    panel3 = null;
    header = null;
    header2 = null;
  });

  it('renders correctly', async () => {
    expect(accordion.outerHTML).toMatchSnapshot();
    panel.expanded = true;
    expect(accordion.outerHTML).toMatchSnapshot();
    panel.expanded = false;
    expect(accordion.outerHTML).toMatchSnapshot();
  });

  it('renders with no errors', async () => {
    const errors = jest.spyOn(global.console, 'error');
    accordion.remove();
    accordion = await createAccordion(accordion);

    expect(document.querySelectorAll('ids-accordion').length).toEqual(1);
    expect(errors).not.toHaveBeenCalled();
  });

  it('can set the pane title attribute', () => {
    const panelTitle = 'Expander text';
    panel.pane.setAttribute('title', panelTitle);
    expect(panel.pane.getAttribute('title')).toBe(panelTitle);
  });

  it('can change its expanded property', () => {
    const panelEl = accordion.querySelector('ids-accordion-panel');

    panelEl.setAttribute('expanded', true);
    panel.expanded = true;
    expect(panelEl.getAttribute('expanded')).toBeTruthy();
    expect(panel.getAttribute('expanded')).toBeTruthy();

    panelEl.setAttribute('expanded', false);
    panel.expanded = false;
    expect(panelEl.getAttribute('expanded')).toBeFalsy();
    expect(panel.expanded).toBeFalsy();
  });

  it('can change set its aria-expanded attribute', () => {
    panel.expanded = true;
    expect(header.getAttribute('aria-expanded')).toBeTruthy();
  });

  it('can be expanded/collapsed when clicked (mouse)', () => {
    const args: any = {
      target: panel.expander,
      bubbles: true,
      cancelable: true,
      view: window
    };
    const event = new MouseEvent('click', args);

    // Expand
    panel.expander.dispatchEvent(event);
    expect(panel.expanded).toBeTruthy();

    // Collapse
    panel.expander.dispatchEvent(event);
    expect(panel.expanded).toBeFalsy();
  });

  it('can be expanded/collapsed when touched', () => {
    const args: any = {
      touches: [{
        pageX: 0,
        pageY: 0,
        target: panel.expander
      }],
      bubbles: true,
      cancpanelable: true,
      view: window
    };
    const event: any = new TouchEvent('touchend', args);

    // Expand
    panel.expander.dispatchEvent(event);

    expect(panel.expanded).toBeTruthy();

    // Collapse
    panel.expander.dispatchEvent(event);

    expect(panel.expanded).toBeFalsy();
  });

  it('can be expanded/collapsed when pressing Enter key', () => {
    const event = new KeyboardEvent('keydown', { key: 'Enter' });

    // Expand
    panel.dispatchEvent(event);
    expect(panel.expanded).toBeTruthy();

    // Collapse
    panel.dispatchEvent(event);
    expect(panel.expanded).toBeFalsy();
  });

  it('can be expanded/collapsed when pressing Space key', () => {
    const event = new KeyboardEvent('keydown', { key: ' ' });

    // Expand
    panel.dispatchEvent(event);
    expect(panel.expanded).toBeTruthy();

    // Collapse
    panel.dispatchEvent(event);
    expect(panel.expanded).toBeFalsy();
  });

  it('can be expanded/collapsed programmatically', async () => {
    // Expand
    header.expanded = true;
    await waitFor(() => expect(panel.expanded).toEqual(true));
    await waitFor(() => expect(header.expanded).toEqual(panel.expanded));

    // Collapse
    header.expanded = false;
    await waitFor(() => expect(panel.expanded).toEqual(false));
    await waitFor(() => expect(header.expanded).toEqual(panel.expanded));
  });

  it('can select the next panel when pressing the ArrowDown key', () => {
    const event = new KeyboardEvent('keydown', { key: 'ArrowDown' });
    let nextPanel = panel.nextElementSibling;

    panel.dispatchEvent(event);
    nextPanel.setAttribute('tabindex', '0');
    expect(nextPanel.getAttribute('tabindex')).toBe('0');

    nextPanel = panel3.nextElementSibling;
    panel3.dispatchEvent(event);
    expect(nextPanel).toBe(null);
  });

  it('can select the prev panel when pressing the ArrowUp key', () => {
    const event = new KeyboardEvent('keydown', { key: 'ArrowUp' });
    let prevPanel = panel2.previousElementSibling;

    panel2.dispatchEvent(event);
    prevPanel.setAttribute('tabindex', '0');
    expect(prevPanel.getAttribute('tabindex')).toBe('0');

    prevPanel = panel.previousElementSibling;
    panel.dispatchEvent(event);
    expect(prevPanel).toBe(null);
  });

  it('can change the height of the pane', async () => {
    panel.pane.style.height = `100px`;

    requestAnimationFrame(() => {
      panel.pane.style.height = `100px`;
      requestAnimationFrame(() => {
        panel.pane.style.height = `0px`;
      });
    });
    await waitFor(() => expect(panel.pane.style.height).toEqual('0px'));
  });

  it('supports setting mode', () => {
    accordion.mode = 'dark';
    expect(accordion.container.getAttribute('mode')).toEqual('dark');
  });

  it('supports setting version', () => {
    accordion.version = 'classic';
    expect(accordion.container.getAttribute('version')).toEqual('classic');
  });

  it('supports setting allow one pane', () => {
    accordion.allowOnePane = true;
    expect(accordion.allowOnePane).toBeTruthy();

    accordion.allowOnePane = false;
    expect(accordion.allowOnePane).toBeFalsy();
  });

  it('supports color variants', async () => {
    accordion = await createAccordion(accordion, 'app-menu');
    expect(accordion.colorVariant).toBe('app-menu');
  });

  it('has a reference to its panels', () => {
    expect(accordion.panels.length).toBe(3);
    expect(accordion.panels.includes(panel3));
  });

  it('should have reference to accordion in accordion-panels', () => {
    expect(panel.accordion).toEqual(accordion);
  });

  it('will not error if no pane', () => {
    panel.container.querySelector('.ids-accordion-pane').remove();
    panel.collapsePane();
    expect(panel.pane).toBe(null);
  });

  it('has a reference to its focused panel', () => {
    panel2.focus();

    expect((document.activeElement as any).isEqualNode(header2)).toBeTruthy();
    expect(accordion.focused.isEqualNode(panel2)).toBeTruthy();

    // Create another element outside the app menu and focus it
    const extraneousElem = document.createElement('input');
    document.body.appendChild(extraneousElem);
    extraneousElem.focus();

    // If no element inside the menu is focused, the property should be undefined
    expect(accordion.focused).toBeFalsy();
  });

  it('can navigate among its panels programatically', () => {
    panel.focus();
    const next = accordion.navigate(1);
    expect(next.isEqualNode(panel2)).toBeTruthy();

    const prev = accordion.navigate(-1);
    expect(prev.isEqualNode(panel)).toBeTruthy();

    // If "0" is passed, stay put
    const none = accordion.navigate(0);
    expect(none.isEqualNode(panel)).toBeTruthy();

    // Don't accept junk values
    const junk = accordion.navigate('junk');
    expect(junk.isEqualNode(panel)).toBeTruthy();

    const noArgs = accordion.navigate();
    expect(noArgs.isEqualNode(panel)).toBeTruthy();
  });

  it('can navigate among its panels using the keyboard', () => {
    const navigateUpEvent = new KeyboardEvent('keydown', { key: 'ArrowUp' });
    const navigateDownEvent = new KeyboardEvent('keydown', { key: 'ArrowDown' });

    // Focus the first one
    panel.focus();
    expect(accordion.focused).toEqual(panel);

    accordion.dispatchEvent(navigateDownEvent);
    expect(accordion.focused).toEqual(panel2);

    accordion.dispatchEvent(navigateUpEvent);
    expect(accordion.focused).toEqual(panel);
  });

  it('has headers that are aware of their expanded status', async () => {
    panel.expanded = true;
    await waitFor(() => expect(header.expanded).toBeTruthy());
  });

  it('can select headers', () => {
    header.selected = true;

    expect(header.container.classList.contains('selected')).toBeTruthy();

    header2.selected = true;

    expect(header.selected).toBeFalsy();
    expect(header2.container.classList.contains('selected')).toBeTruthy();
  });

  it('can change its headers expander type', async () => {
    await waitFor(() => expect(header.expanderType).toBe('caret'));

    header.expanderType = 'plus-minus';
    await waitFor(() => expect(header.expanderType).toBe('plus-minus'));
  });

  it('can add/remove icons from accordion headers', () => {
    const icon = header.container.querySelector('.ids-accordion-display-icon');
    header.icon = 'user';

    expect(header.getAttribute('icon')).toBe('user');
    expect(icon.icon).toBe('user');

    header.icon = null;

    expect(header.getAttribute('icon')).toBe(null);
    expect(icon.icon).toBe(null);
  });

  it('toggle expander icon for header', async () => {
    const icon = header.container.querySelector('.ids-accordion-expander-icon');
    expect(icon.getAttribute('icon')).toBe('caret-down');
    header.toggleExpanderIcon(true);
    await waitFor(() => expect(icon.getAttribute('icon')).toBe('caret-down'));
  });

  it('should update with container language change', () => {
    const container: any = new IdsContainer();
    document.body.appendChild(container);
    container.appendChild(accordion);

    const language = { before: 'en', after: 'ar' };
    const mockCallback = jest.fn((e) => {
      expect(e.detail.language.name).toEqual(language.after);
    });

    expect(accordion.language.name).toEqual(language.before);
    container.addEventListener('languagechange', mockCallback);
    const event = new CustomEvent('languagechange', {
      detail: { language: { name: language.after } }
    });
    container.dispatchEvent(event);

    expect(mockCallback.mock.calls.length).toBe(1);
  });

  it('should have only one pane expanded with allowOnePane setting', () => {
    accordion.allowOnePane = true;
    panel.expanded = true;
    panel2.expanded = true;

    expect(panel.expanded).toBeFalsy();
    expect(panel2.expanded).toBeTruthy();
    expect(panel3.expanded).toBeFalsy();
  });

  it('should add/remove alignment classes with panel\'s contentAlignment', () => {
    panel.contentAlignment = '';
    expect(header.container.classList.contains('has-icon')).toBeFalsy();

    panel.contentAlignment = 'has-icon';
    expect(header.container.classList.contains('has-icon')).toBeTruthy();

    panel.contentAlignment = 'has-menu';
    expect(header.container.classList.contains('has-icon')).toBeFalsy();
  });

  it('should not expand when panel is disabled', () => {
    panel.disabled = true;
    const event = new KeyboardEvent('keydown', { key: ' ' });

    panel.dispatchEvent(event);
    expect(panel.expanded).toBeFalsy();

    panel.disabled = false;
    // Expand
    panel.dispatchEvent(event);
    expect(panel.expanded).toBeTruthy();
  });

  it('should not expand all panels when accordion disabled', () => {
    accordion.disabled = true;
    const event = new KeyboardEvent('keydown', { key: ' ' });

    panel.dispatchEvent(event);
    expect(panel.expanded).toBeFalsy();

    panel2.dispatchEvent(event);
    expect(panel2.expanded).toBeFalsy();

    panel2.dispatchEvent(event);
    expect(panel2.expanded).toBeFalsy();
  });
});
