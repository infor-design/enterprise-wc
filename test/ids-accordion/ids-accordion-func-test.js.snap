// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdsAccordion Component renders correctly 1`] = `
"<ids-accordion language=\\"en\\" locale=\\"en-US\\">
    <ids-accordion-panel id=\\"p1\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h1\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 1</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p2\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h2\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 2</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p3\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h3\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 3</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
  </ids-accordion>"
`;

exports[`IdsAccordion Component renders correctly 2`] = `
"<ids-accordion language=\\"en\\" locale=\\"en-US\\">
    <ids-accordion-panel id=\\"p1\\" language=\\"en\\" locale=\\"en-US\\" expanded=\\"true\\">
      <ids-accordion-header id=\\"h1\\" slot=\\"header\\" aria-expanded=\\"true\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 1</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p2\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h2\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 2</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p3\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h3\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 3</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
  </ids-accordion>"
`;

exports[`IdsAccordion Component renders correctly 3`] = `
"<ids-accordion language=\\"en\\" locale=\\"en-US\\">
    <ids-accordion-panel id=\\"p1\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h1\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 1</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p2\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h2\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 2</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
    <ids-accordion-panel id=\\"p3\\" language=\\"en\\" locale=\\"en-US\\">
      <ids-accordion-header id=\\"h3\\" slot=\\"header\\" aria-expanded=\\"false\\" language=\\"undefined\\" locale=\\"en-US\\">
        <ids-text>Header 3</ids-text>
      </ids-accordion-header>
    </ids-accordion-panel>
  </ids-accordion>"
`;
